### ISSUE TITLE
pyc files for build dependencies being packaged

### ISSUE THREAD

**Issue by rmcgibbo** [conda/conda-build#718]
Hi,

I just looked at the contents of https://anaconda.org/omnia/mdtraj/1.5.1/download/win-32/mdtraj-1.5.1-np110py35_0.tar.bz2, (compiled using this recipe about a month ago: https://github.com/omnia-md/conda-recipes/blob/master/mdtraj/meta.yaml), and noticed an odd issue.

The package contains `.pyc` files not only for the project, but also for `numpy` and `cython`, both of which are build-time dependencies. I guess when the package is built and numpy and Cython are imported, that triggers the generation of those packages' `.pyc` files which conda-build then interprets incorrectly. as being build artifacts of the current project.

I guess this issue is similar to https://github.com/conda/conda-build/pull/715, not also kind of different.

**Comment by rmcgibbo** [comment]
```
$ tar -tf mdtraj-1.5.1-np110py35_0.tar.bz2
Lib/site-packages/Cython/Build/__pycache__/Dependencies.cpython-35.pyc
Lib/site-packages/Cython/Build/__pycache__/__init__.cpython-35.pyc
Lib/site-packages/Cython/Compiler/__pycache__/Annotate.cpython-35.pyc
Lib/site-packages/Cython/Compiler/__pycache__/Buffer.cpython-35.pyc
[... etc ...]
```

**Comment by ivoflipse** [comment]
I encountered a similar problem where my CI build failed, because it was compiling .pyc files for a dependency, which happened to include Python 3 code that never should get called from Python 2. I ended up creating a conda package for that project that simply excluded the Python 3 code from the Python 2 package, but it would be great if I didn't have to resort to such workarounds

**Comment by tacaswell** [comment]
possibly related: https://github.com/conda/conda-build/issues/369

**Comment by ivoflipse** [comment]
Likewise this one: https://github.com/conda/conda-build/pull/370

**Comment by meawoppl** [comment]
I had a problem in PIL (pillow) that may be related as well.

**Comment by pelson** [comment]
Bump. This is really not healthy for conda-build.

**Comment by kalefranz** [comment]
Agreed. @csoja can you help prioritize this on @ilanschnell's list. I think he's best equipped to take the first stab.

**Comment by kalefranz** [comment]
For the example that @rmcgibbo gave, is there a reason I'm missing why the numpy and cython packages don't include their own pyc files? Were they just packaged incorrectly?

**Comment by rmcgibbo** [comment]
¯\_(ツ)_/¯

**Comment by pelson** [comment]
Seems this issue is bleeding into recipes https://github.com/conda/conda-recipes/blob/master/python/pillow/build.sh#L12. Has this been prioritised?

**Comment by Korijn** [comment]
We're being affected by #369 again. Please prioritise?

**Comment by ilanschnell** [comment]
I would argue that is this is not really a conda-build bug, but rather a bug in:
  * conda packages which the recipe depends on at build time, because they are missing `.pyc` which then get created during the build process.
  * the recipe itself, because it is always possible to remove the excess `.pyc` during the build, e.g. you can add something like `rm -rf $SP_DIR/Cython` right after `$PYTHON setup.py install` in `build.sh`.

What would a fix to conda-build look like?  How can conda-build generally know these creation of these `.pyc` files is not intentional?

**Comment by Korijn** [comment]
Fair point. So how do we go about fixing the PyQt4 package from the defaults channel?

**Comment by msarahan** [comment]
I think it's a good idea to try to have conda build not do strange things
here, but I am not sure how to approach it. We can't exclude any pyc
without a matching py, because then people could not make packages with pyc
but no py files.

What about basing it off of the files in the source? Unless people are
dynamically creating py files at install time, that might work.

**Comment by marscher** [comment]
you could export PYTHONDONTWRITEBYTECODE=0 somewhere before the actual build process is started to avoid generation extra pyc files. This should be a minimal invasive approach.

**Comment by jakirkham** [comment]
> you could export PYTHONDONTWRITEBYTECODE=0 somewhere before the actual build process is started to avoid generation extra pyc files. This should be a minimal invasive approach.

That's not a bad idea, but what guarantees do we have that `compile_missing_pyc` won't compile them for us?

**Comment by jakirkham** [comment]
This was fixed around `conda-build` 2 IIRC. Should be closed then.
